src/TorchSharp/DisposeScope.cs

```diff
@@ -0,0 +1,301 @@
+﻿// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using TorchSharp.Utils;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Keeps track of all disposables that are in the current scope - the dispose scopes can be nested and the
+    /// nesting functionality is mainly managed by DisposeScopeManager.
+    /// </summary>
+    public class DisposeScope : IDisposable
+    {
+        private readonly DisposeScopeManager _disposeScopeManager;
+
+        public DisposeScope(DisposeScopeManager disposeScopeManager)
```

Comment:
Does it make sense to have a public `DisposeScope` constructor? Shouldn't one always create a `DisposeScope` from a manager (which should also push it to the stack?)?

Reply:
True, I'm not sure how fixes like the one you're suggesting is currently managed, would I start a new PR?

Reply:
@mfagerlund:

>> I'm not sure how fixes like the one you're suggesting is currently managed, would I start a new PR?

Yes. Make sure to merge everything from this repo's main before starting new work, just to make sure that you have everything that's gone in.